subprojects/execution/src/main/java/org/gradle/internal/execution/history/impl/FingerprintMapSerializer.java

```diff
@@ -50,13 +50,13 @@ public FingerprintMapSerializer(StringInterner stringInterner) {
     @Override
     public Map<String, FileSystemLocationFingerprint> read(Decoder decoder) throws IOException {
         int fingerprintCount = decoder.readSmallInt();
-        Map<String, FileSystemLocationFingerprint> fingerprints = new LinkedHashMap<String, FileSystemLocationFingerprint>(fingerprintCount);
+        ImmutableMap.Builder<String, FileSystemLocationFingerprint> fingerprints = ImmutableMap.builderWithExpectedSize(fingerprintCount);
         for (int i = 0; i < fingerprintCount; i++) {
             String absolutePath = stringInterner.intern(decoder.readString());
             FileSystemLocationFingerprint fingerprint = readFingerprint(decoder);
             fingerprints.put(absolutePath, fingerprint);
         }
-        return fingerprints;
+        return fingerprints.build();
```

Comment:
Cool! How did you measure that this is more efficient?

Reply:
I didn't, I just know that an immutable map is always smaller than a linked one, since the latter needs to create a node object for each entry while the former just uses an array.

Reply:
Technically speaking, https://openjdk.java.net/projects/code-tools/jol/ can be used.
There's `GraphLayout.parseInstance(m).totalSize()` API there.

Here are my rants:

`new LinkedHashMap<String, String>((int) ( j / 0.75), 1.0f);`

Note: the sizes below are pure overhead (it does not include the stored keys and values)

1.8.0_152
```
elements;LinkedSize;ImmutableSize;PerEntryDiff;TotalDiff
0;72;0;72;72
1;192;96;96;96
2;288;248;20;40
3;392;344;16;48
4;504;424;20;80
5;600;528;14;72
6;696;608;15;88
7;824;696;18;128
8;920;776;18;144
9;1016;864;17;152
10;1112;976;14;136
11;1208;1064;13;144
12;1304;1144;13;160
13;1464;1232;18;232
14;1560;1312;18;248
15;1656;1400;17;256
16;1744;1472;17;272
17;1848;1568;16;280
18;1944;1648;16;296
19;2040;1736;16;304
20;2128;1872;13;256
21;2224;1960;13;264
22;2320;2040;13;280
23;2416;2128;13;288
24;2520;2216;13;304
25;2744;2304;18;440
26;2840;2384;18;456
27;2928;2464;17;464
28;3032;2552;17;480
29;3120;2632;17;488
30;3224;2720;17;504
31;3320;2808;17;512
32;3408;2880;17;528
33;3512;2976;16;536
34;3608;3056;16;552
35;3696;3136;16;560
36;3800;3224;16;576
37;3880;3296;16;584
38;3992;3392;16;600
39;4088;3608;12;480
40;4184;3688;12;496
41;4280;3776;12;504
42;4376;3856;12;520
43;4472;3944;12;528
44;4560;4016;12;544
45;4656;4104;12;552
46;4760;4192;12;568
47;4856;4280;12;576
48;4944;4352;12;592
49;5304;4448;17;856
50;5392;4520;17;872
51;5480;4600;17;880
52;5592;4696;17;896
53;5680;4776;17;904
54;5784;4864;17;920
55;5880;4952;17;928
56;5976;5032;17;944
57;6072;5120;17;952
58;6160;5192;17;968
59;6256;5280;17;976
60;6352;5360;17;992
61;6456;5456;16;1000
62;6552;5536;16;1016
63;6648;5624;16;1024
64;6736;5696;16;1040
65;6832;5784;16;1048
66;6928;5864;16;1064
67;7024;5952;16;1072
68;7128;6040;16;1088
69;7208;6112;16;1096
70;7296;6184;16;1112
71;7400;6280;16;1120
72;7512;6376;16;1136
73;7592;6448;16;1144
74;7688;6528;16;1160
75;7792;6624;16;1168
76;7896;6712;16;1184
77;7984;7048;12;936
78;8072;7120;12;952
79;8168;7208;12;960
80;8280;7304;12;976
81;8376;7392;12;984
82;8472;7472;12;1000
83;8552;7544;12;1008
84;8664;7640;12;1024
85;8752;7720;12;1032
86;8856;7808;12;1048
87;8952;7896;12;1056
88;9040;7968;12;1072
89;9136;8056;12;1080
90;9240;8144;12;1096
91;9336;8232;12;1104
92;9424;8304;12;1120
93;9528;8400;12;1128
94;9608;8464;12;1144
95;9696;8544;12;1152
96;9808;8640;12;1168
97;10416;8728;17;1688
98;10520;8816;17;1704
99;10608;8896;17;1712
100;10696;8968;17;1728
```

Java 13:

```
0;72;0;72;72
1;184;88;96;96
2;272;232;20;40
3;368;320;16;48
4;472;392;20;80
5;560;488;14;72
6;648;560;15;88
7;768;640;18;128
8;856;712;18;144
9;944;792;17;152
10;1032;896;14;136
11;1120;976;13;144
12;1208;1048;13;160
13;1360;1128;18;232
14;1448;1200;18;248
15;1536;1280;17;256
16;1624;1352;17;272
17;1712;1432;16;280
18;1800;1504;16;296
19;1888;1584;16;304
20;1976;1720;13;256
21;2064;1800;13;264
22;2152;1872;13;280
23;2240;1952;13;288
24;2328;2024;13;304
25;2544;2104;18;440
26;2632;2176;18;456
27;2720;2256;17;464
28;2808;2328;17;480
29;2896;2408;17;488
30;2984;2480;17;504
31;3072;2560;17;512
32;3160;2632;17;528
33;3248;2712;16;536
34;3336;2784;16;552
35;3424;2864;16;560
36;3512;2936;16;576
37;3600;3016;16;584
38;3688;3088;16;600
39;3776;3296;12;480
40;3864;3368;12;496
41;3952;3448;12;504
42;4040;3520;12;520
43;4128;3600;12;528
44;4216;3672;12;544
45;4304;3752;12;552
46;4392;3824;12;568
47;4480;3904;12;576
48;4568;3976;12;592
49;4912;4056;17;856
50;5000;4128;17;872
51;5088;4208;17;880
52;5176;4280;17;896
53;5264;4360;17;904
54;5352;4432;17;920
55;5440;4512;17;928
56;5528;4584;17;944
57;5616;4664;17;952
58;5704;4736;17;968
59;5792;4816;17;976
60;5880;4888;17;992
61;5968;4968;16;1000
62;6056;5040;16;1016
63;6144;5120;16;1024
64;6232;5192;16;1040
65;6320;5272;16;1048
66;6408;5344;16;1064
67;6496;5424;16;1072
68;6584;5496;16;1088
69;6672;5576;16;1096
70;6760;5648;16;1112
71;6848;5728;16;1120
72;6936;5800;16;1136
73;7024;5880;16;1144
74;7112;5952;16;1160
75;7200;6032;16;1168
76;7288;6104;16;1184
77;7376;6440;12;936
78;7464;6512;12;952
79;7552;6592;12;960
80;7640;6664;12;976
81;7728;6744;12;984
82;7816;6816;12;1000
83;7904;6896;12;1008
84;7992;6968;12;1024
85;8080;7048;12;1032
86;8168;7120;12;1048
87;8256;7200;12;1056
88;8344;7272;12;1072
89;8432;7352;12;1080
90;8520;7424;12;1096
91;8608;7504;12;1104
92;8696;7576;12;1120
93;8784;7656;12;1128
94;8872;7728;12;1144
95;8960;7808;12;1152
96;9048;7880;12;1168
97;9648;7960;17;1688
98;9736;8032;17;1704
99;9824;8112;17;1712
100;9912;8184;17;1728
```

`new HashMap<String, String>((int) ( j / 0.75), 1.0f)`
Java 1.8.0_152.
Note: it is not directly comparable because it shuffles the iteration order.

```
0;64;0;64;64
1;176;96;80;80
2;264;248;8;16
3;360;344;5;16
4;464;424;10;40
5;552;528;5;24
6;640;608;5;32
7;760;696;9;64
8;848;776;9;72
9;936;864;8;72
10;1024;976;5;48
11;1112;1064;4;48
12;1200;1144;5;56
13;1352;1232;9;120
14;1440;1312;9;128
15;1528;1400;9;128
16;1608;1472;9;136
17;1704;1568;8;136
18;1792;1648;8;144
19;1880;1736;8;144
20;1960;1872;4;88
21;2048;1960;4;88
22;2136;2040;4;96
23;2224;2128;4;96
24;2320;2216;4;104
25;2536;2304;9;232
26;2624;2384;9;240
27;2704;2464;9;240
28;2800;2552;9;248
29;2880;2632;9;248
30;2976;2720;9;256
31;3064;2808;8;256
32;3144;2880;8;264
33;3240;2976;8;264
34;3328;3056;8;272
35;3408;3136;8;272
36;3504;3224;8;280
37;3576;3296;8;280
38;3680;3392;8;288
39;3768;3608;4;160
40;3856;3688;4;168
41;3944;3776;4;168
42;4032;3856;4;176
43;4120;3944;4;176
44;4200;4016;4;184
45;4288;4104;4;184
46;4384;4192;4;192
47;4472;4280;4;192
48;4552;4352;4;200
49;4904;4448;9;456
50;4984;4520;9;464
51;5064;4600;9;464
52;5168;4696;9;472
53;5248;4776;9;472
54;5344;4864;9;480
55;5432;4952;9;480
56;5520;5032;9;488
57;5608;5120;9;488
58;5688;5192;9;496
59;5776;5280;8;496
60;5864;5360;8;504
61;5960;5456;8;504
62;6048;5536;8;512
63;6136;5624;8;512
64;6216;5696;8;520
65;6304;5784;8;520
66;6392;5864;8;528
67;6480;5952;8;528
68;6576;6040;8;536
69;6648;6112;8;536
70;6728;6184;8;544
71;6824;6280;8;544
72;6928;6376;8;552
73;7000;6448;8;552
74;7088;6528;8;560
75;7184;6624;7;560
76;7280;6712;7;568
77;7360;7048;4;312
78;7440;7120;4;320
79;7528;7208;4;320
80;7632;7304;4;328
81;7720;7392;4;328
82;7808;7472;4;336
83;7880;7544;4;336
84;7984;7640;4;344
85;8064;7720;4;344
86;8160;7808;4;352
87;8248;7896;4;352
88;8328;7968;4;360
89;8416;8056;4;360
90;8512;8144;4;368
91;8600;8232;4;368
92;8680;8304;4;376
93;8776;8400;4;376
94;8848;8464;4;384
95;8928;8544;4;384
96;9032;8640;4;392
97;9632;8728;9;904
98;9728;8816;9;912
99;9808;8896;9;912
100;9888;8968;9;920
```

Reply:
The test code (it is mine, and it can be used under the terms of Apache 2.0):

```java
import com.google.common.collect.ImmutableMap;
import org.junit.jupiter.api.Test;
import org.openjdk.jol.info.GraphLayout;

import java.lang.reflect.Field;
import java.util.*;

public class CompactMapSizeTest {
    private final static ArrayList<String> keys = new ArrayList<String>();

    private final BitSet seenKeys = new BitSet();

    static {
        for (int i = 0; i < 1000; i++) {
            keys.add("key" + i);
        }
    }

    private int fill(Map<String, String> map, int size, Random rnd, boolean keySize, boolean valueSize) {
        int overhead = 0;
        seenKeys.clear();
        for (int i = 0; i < size; i++) {
            int index = rnd.nextInt(keys.size());
            if (seenKeys.get(index)) {
                i--;
                continue;
            }
            seenKeys.set(index);
            String key = keys.get(index);
            String value = "value" + index;
            map.put(key, value);
            if (keySize)
                overhead += GraphLayout.parseInstance(key).totalSize();
            if (valueSize)
                overhead += GraphLayout.parseInstance(value).totalSize();
        }
        return overhead;
    }

    @Test
    public void mapSizeOverhead() throws NoSuchFieldException, IllegalAccessException {
        Random rnd = new Random();
        Field klass = CompactHashMap.class.getDeclaredField("klass");
        klass.setAccessible(true);

        System.out.println("elements;LinkedSize;ImmutableSize;PerEntryDiff;TotalDiff");
        for (int j = 0; j < 101; j++) {
            rnd.setSeed(42 * j);
            Map<String, String> m = new HashMap<String, String>((int) ( j / 0.75), 1.0f);
            long valuesSize = fill(m, j, rnd, true, true);

            rnd.setSeed(42 * j);
            Map<Object, Object> im = ImmutableMap.builderWithExpectedSize(j)
                    .putAll(m)
                    .build();

            long linkedTotalSize = GraphLayout.parseInstance(m).totalSize();

            long immutableTotalSize = GraphLayout.parseInstance(im).totalSize();
            if (j == 0) { immutableTotalSize = 0; }
            System.out.println(j
                    + ";" + (linkedTotalSize - valuesSize)
                    + ";" + (immutableTotalSize - valuesSize)
                    + ";" + Math.round(j==0?(linkedTotalSize - immutableTotalSize):(linkedTotalSize - immutableTotalSize)*1.0/j)
                    + ";" + (linkedTotalSize - immutableTotalSize)
            );
        }
    }
}
```

Reply:
This is really useful @vlsi, thanks!